README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->
<a href="https://github.com/inf0rmatix/design_grid/actions">
    <img src="https://github.com/inf0rmatix/design_grid/workflows/tests/badge.svg" alt="Build Status">
</a> 
<a href="https://codecov.io/gh/inf0rmatix/design_grid" > 
 <img src="https://codecov.io/gh/inf0rmatix/design_grid/branch/master/graph/badge.svg?token=7OWGIYJD5Y"/> 
</a>


Implement designs just like in Figma. Arrange your widgets in a grid and make your designer happy.

Design Grid streamlines your workflow and saves time. Easily configure multiple devices and screen sizes while reducing boilerplate code. Its user-friendly design makes integration effortless, allowing you to focus on your project's core.

The classic use-case of implementing a colum-based design. Use ResponsiveDesignGrid to implement a bootstrap-like design system where every element has a different column-span based on breakpoints.

Need to implement a design based on the strict rules of Material Layout? Use the MaterialDesignGrid to get an exact implementation of the [Material Design Layout guide](https://m2.material.io/design/layout/responsive-layout-grid.html).

## 🔗 Table of Contents
- [🔗 Table of Contents](#-table-of-contents)
- [✨ Features](#-features)
  - [Responsive Design Grid](#responsive-design-grid)
  - [Material Design Grid](#material-design-grid)
- [🎉 Getting started](#-getting-started)
- [🏗️ Usage](#️-usage)
  - [Responsive Design Grid](#responsive-design-grid-1)
    - [Basic Usage](#basic-usage)
    - [Adjusting Breakpoints](#adjusting-breakpoints)
    - [Customizing the grid](#customizing-the-grid)
  - [Material Design Grid](#material-design-grid-1)
- [📚 Additional information](#-additional-information)
  - [Implementing a custom design grid approach](#implementing-a-custom-design-grid-approach)


## ✨ Features

### Responsive Design Grid

Implement a column-based design with different column-spans for every breakpoint.
Responsive design grid widgets can be indefinitely nested.
Create small and clean widgets that use ResponsiveDesignGrid for layout.
They will fit, no matter how deeply nested you use them.

You can customize:

- Breakpoints
- Column-spacing
- Row-spacing
- Grid-padding
- Amount of columns

![](https://github.com/inf0rmatix/design_grid/blob/bafda9b5bdaee878da73c6407115d247872dc3a8/readme/responsive_design_grid.gif)

### Material Design Grid

Implement a design based on the [Material Design guidelines](https://m2.material.io/design/layout/responsive-layout-grid.html#columns-gutters-and-margins).
Material Design Grid is not meant to be used nested.

You can customize:

- Breakpoints
- Column-spacing
- Row-spacing
- Grid-padding
- Amount of columns per breakpoint

![](https://github.com/inf0rmatix/design_grid/blob/bafda9b5bdaee878da73c6407115d247872dc3a8/readme/material_design_grid.gif)

## 🎉 Getting started

Decide whether you need to use the `ResponsiveDesignGrid` or the `MaterialDesignGrid`.
Speak with your designer to build a common understanding.

## 🏗️ Usage

### Responsive Design Grid

`ResponsiveDesignGridConfig` is required only once in your app, it provides the `ResponsiveDesignGridTheme`.
You can however always add an extra `ResponsiveDesignGridTheme` widget if you really need to. Beware, that this is not recommended.

You can nest design grids indefinitely. To keep performance high, make sure to nest them directly in a `ResponsiveDesignGridItem`.
If you get any errors, you might have to pass the parameter `shouldCalculateLayout: true` to the `ResponsiveDesignGrid` widget. 
This might happen most likely when the design grid does not get the full width of the app, or if you nested it inside another `ResponsiveDesignGridItem` and it doesn't get the full width there.

`ResponsiveDesignGridRow` widget will wrap its children, your rows should have as little children as possible. Avoid large Rows since this will impact performance.

#### Basic Usage

```dart
@override
Widget build(BuildContext context){
    return ResponsiveDesignGridConfig(
        child: ResponsiveDesignGrid(
            children: [
                ResponsiveDesignGridRow(
                    children: [
                        ResponsiveDesignGridItem(
                            columns: ResponsiveDesignGridColumns(
                                small: 12,
                                medium: 10,
                                large: 8,
                                extraLarge: 6,
                            ),
                            child: Placeholder(),
                        ),
                    ],
                )
            ],
        ),
    );
}
```

#### Adjusting Breakpoints

To change the default breakpoints values, you can use the `breakpoints` property in `ResponsiveDesignGridConfig` constructor which takes an instance of `ResponsiveDesignGridBreakpoints` class.


```dart
@override
Widget build(BuildContext context){
    return ResponsiveDesignGridConfig(
            breakpoints: const ResponsiveDesignGridBreakpoints(
            small: 375,
            medium: 768,
            large: 1280,
            extraLarge: 1920,
        ),
        child: ResponsiveDesignGrid(
            children: [
                ResponsiveDesignGridRow(
                    children: [
                        ResponsiveDesignGridItem(
                            columns: ResponsiveDesignGridColumns(
                                small: 12,
                                medium: 10,
                                large: 8,
                                extraLarge: 6,
                            ),
                            child: Placeholder(),
                        ),
                    ],
                )
            ],
        ),
    );
}
```


#### Customizing the grid

To change the amount of columns, the column spacing, the row spacing or the grid padding, you can use the `theme` property in `ResponsiveDesignGridConfig` constructor which takes an instance of `ResponsiveDesignGridThemeData` class.

```dart
@override
Widget build(BuildContext context){
    return ResponsiveDesignGridConfig(
        theme: ResponsiveDesignGridThemeData(
            columns: 12,
            columnSpacing: 16.0,
            rowSpacing: 16.0,
            gridPadding:16.0,
        ),
        child: ResponsiveDesignGrid(
            children: [
                ResponsiveDesignGridRow(
                    children: [
                        ResponsiveDesignGridItem(
                            columns: ResponsiveDesignGridColumns(
                                small: 12,
                                medium: 10,
                                large: 8,
                                extraLarge: 6,
                            ),
                            child: Placeholder(),
                        ),
                    ],
                )
            ],
        ),
    );
}
```


### Material Design Grid

`MaterialDesignGridTheme` is required only once in your app, above the `MaterialDesignGrid` widget.

`MaterialDesignGrid` is **not** meant to be used nested due to material specifications.

`MaterialDesignGridRow` widget will wrap its children, you should size your rows as small as possible. Avoid large Rows since this will impact performance.

```dart
@override
Widget build(BuildContext context) {
    return MaterialDesignGridTheme(
        child: MaterialDesignGrid(
            children: [
                MaterialDesignGridRow(
                    children: [
                        MaterialDesignGridItem(
                            columns: 4,
                            child: Placeholder(),
                        ),
                    ],
                )
            ],
        ),
    );
}
```

## 📚 Additional information

All classes and exposed properties are well documented. If you find the documentation to be lacking, please [create an issue](https://github.com/inf0rmatix/design_grid/issues/new/choose) 😊


### Implementing a custom design grid approach

The base widgets used for both the `ResponsiveDesignGrid` as well as the `MaterialDesignGrid` can be found in `lib/src/widgets`.
You can extend those to create your super-fancy very own custom design grid implementation 🥳.


CHANGELOG.md
## 1.1.0

 * Implementing `DesignGridLayoutType` to enable switching between column and list view
 * Correcting minor documentation flaws

## 1.0.2

 * adding screenshots

## 1.0.1

 * adding repository url to pubspec.yaml

## 1.0.0

 * fixing README.md

## 0.1.1

 * removing old property "breakpoints" from ResponsiveDesignGridThemeData
  
## 0.1.0

 * updating README.md

## 0.0.4

 * fixing gif links in readme

## 0.0.3

 * Updating readme

## 0.0.2

 * Fixing license information

## 0.0.1

* Initial release


example/lib/main.dart
import 'package:design_grid/design_grid.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Brightness brightness = Brightness.light;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Design Grid Example',
      theme: ThemeData(
        useMaterial3: true,
        colorScheme: ColorScheme.fromSeed(
          seedColor: Colors.blue,
          brightness: brightness,
        ),
      ),
      home: Scaffold(
        appBar: AppBar(
          leading: const Icon(Icons.grid_on_rounded),
          actions: [
            IconButton(
              onPressed: () => setState(() {
                brightness = brightness == Brightness.dark ? Brightness.light : Brightness.dark;
              }),
              icon: brightness == Brightness.dark
                  ? const Icon(Icons.light_mode_rounded)
                  : const Icon(Icons.dark_mode_rounded),
            ),
          ],
          title: const Text('Design Grid Example'),
        ),
        body: ListView(
          padding: const EdgeInsets.symmetric(vertical: 16.0),
          children: const [
            ResponsiveDesignGridExample(),
            Divider(
              height: 64.0,
            ),
            MaterialDesignGridExample(),
          ],
        ),
      ),
    );
  }
}

class ResponsiveDesignGridExample extends StatelessWidget {
  const ResponsiveDesignGridExample({super.key});

  @override
  Widget build(BuildContext context) {
    // The ResponsiveDesignGridConfig widget is used to configure the ResponsiveDesignGrid.
    // This widget should be placed as high up in the widget tree as possible.
    // You can also pass a custom theme to change row spacing, column spacing, the grid outer padding and even the number of columns for the grid.
    return ResponsiveDesignGridConfig(
      breakpoints: const ResponsiveDesignGridBreakpoints(
        small: 375,
        medium: 768,
        large: 1280,
        extraLarge: 1920,
      ),
      // The ResponsiveDesignGridDebugOverlay is a useful widget to visually debug your grid implementation.
      child: ResponsiveDesignGridDebugOverlay(
        // To hide the button in release mode, set enableControls to false. Easiest way is to use kDebugMode.
        enableControls: kDebugMode,
        child: ResponsiveDesignGrid(
          children: [
            // This is a responsive design grid row. Contrary to Flutter's Row widget, this is more a logical separation and will wrap its children if necessary.
            // Keep your rows as small as possible. To enforce a visual line break, prefer using a new Responsive Design Grid Row.
            // For some rare occasions, there is the ResponsiveDesignGridBreak widget which enforces a visual line break within a row.
            ResponsiveDesignGridRow(
              alignment: DesignGridRowAlignment.center,
              children: [
                // This is a responsive design grid item, which spans 12 columns on small screens, 10 columns on medium screens, 8 columns on large screens, and 6 columns on extra large screens.
                ResponsiveDesignGridItem(
                  columns: const ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 8,
                    large: 6,
                    extraLarge: 4,
                  ),
                  // The child widget will be wrapped in a SizedBox widget with constrained width according to the columns property.
                  child: Card(
                    child: Column(
                      children: const [
                        ListTile(
                          title: Text('Responsive Design Grid'),
                        ),
                        Padding(
                          padding: EdgeInsets.all(16.0),
                          child: Text(
                              'This example shows how to use the ResponsiveDesignGrid. This element spans 12 columns on small screens, 10 columns on medium screens, 8 columns on large screens, and 6 columns on extra large screens.\nThe floating action button toggles the debug overlay.'),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
            ResponsiveDesignGridRow(
              alignment: DesignGridRowAlignment.center,
              children: [
                ResponsiveDesignGridItem(
                  columns: const ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 10,
                    large: 8,
                  ),
                  child: TextFormField(
                    decoration: const InputDecoration(
                      labelText: 'Name*',
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
              ],
            ),
            ResponsiveDesignGridRow(
              alignment: DesignGridRowAlignment.center,
              children: [
                ResponsiveDesignGridItem(
                  columns: const ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 5,
                    large: 4,
                  ),
                  child: TextFormField(
                    decoration: const InputDecoration(
                      labelText: 'E-Mail*',
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
                ResponsiveDesignGridItem(
                  columns: const ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 5,
                    large: 4,
                  ),
                  child: TextFormField(
                    decoration: const InputDecoration(
                      labelText: 'Phone (optional)',
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
              ],
            ),
            ResponsiveDesignGridRow(
              alignment: DesignGridRowAlignment.center,
              children: [
                ResponsiveDesignGridItem(
                  columns: const ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 10,
                    large: 8,
                  ),
                  child: TextFormField(
                    minLines: 4,
                    maxLines: 4,
                    decoration: const InputDecoration(
                      labelText: 'Message*',
                      alignLabelWithHint: true,
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
              ],
            ),
            const ResponsiveDesignGridRow(
              alignment: DesignGridRowAlignment.center,
              children: [
                ResponsiveDesignGridItem(
                  columns: ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 10,
                    large: 8,
                  ),
                  child: Text(
                      'Iste aut cupiditate adipisci et iste incidunt est animi. Voluptatem natus reiciendis magni nesciunt magni. Reiciendis ad beatae. Odit corrupti eos voluptatem molestiae voluptatem qui rerum animi officiis. Natus dolores provident excepturi nostrum omnis voluptatem.'),
                ),
              ],
            ),
            ResponsiveDesignGridRow(
              alignment: DesignGridRowAlignment.end,
              children: [
                ResponsiveDesignGridItem(
                  columns: const ResponsiveDesignGridColumns(
                    small: 12,
                    medium: 4,
                    large: 3,
                  ),
                  child: ElevatedButton(
                    onPressed: () {},
                    child: const Text('Submit'),
                  ),
                ),
                ResponsiveDesignGridItem.placeholder(
                  columns: const ResponsiveDesignGridColumns(
                    small: 0,
                    medium: 1,
                    large: 2,
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class MaterialDesignGridExample extends StatelessWidget {
  const MaterialDesignGridExample({super.key});

  @override
  Widget build(BuildContext context) {
    // The MaterialDesignGridTheme widget is used to configure the MaterialDesignGrid.
    // Place this widget as high up in the widget tree as possible.
    // You can also pass a custom theme to change spacings and even custom breakpoints.
    // The breakpoints control how many columns there are, the outer padding and the (optional) maximum body width.
    return MaterialDesignGridTheme(
      // The MaterialDesignGridDebugOverlay is a useful widget to visually debug your grid implementation.
      child: MaterialDesignGridDebugOverlay(
        child: MaterialDesignGrid(
          children: [
            // This is a Material Design Grid Row, contrary to Flutter's Row widget, this one wraps children when necessary.
            // Keep your rows as small as possible. To enforce a visual line break, prefer using a new Material Design Grid Row.
            // For some rare occasions, there is the MaterialDesignGridBreak widget which enforces a visual line break within a row.
            MaterialDesignGridRow(
              children: [
                // Material Design Grid Items are given a fixed column span, which must not be greater then the smallest amount of columns defined in the breakpoints. (By default, this is 4 columns)
                MaterialDesignGridItem(
                  columns: 4,
                  child: Card(
                    child: Column(
                      children: const [
                        ListTile(
                          title: Text('Material Design Grid'),
                        ),
                        Padding(
                          padding: EdgeInsets.all(16.0),
                          child: Text(
                              'This is a Material Design Grid example. The Material Design Grid uses a varying number of columns depending on the screen size. Its children are given a fixed number of columns.\nThe floating action button toggles the debug overlay.'),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
            MaterialDesignGridRow(
              children: [
                MaterialDesignGridItem(
                  columns: 4,
                  child: TextFormField(
                    decoration: const InputDecoration(
                      labelText: 'Name*',
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
              ],
            ),
            MaterialDesignGridRow(
              children: [
                MaterialDesignGridItem(
                  columns: 4,
                  child: TextFormField(
                    decoration: const InputDecoration(
                      labelText: 'E-Mail*',
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
                MaterialDesignGridItem(
                  columns: 4,
                  child: TextFormField(
                    decoration: const InputDecoration(
                      labelText: 'Phone (optional)',
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
              ],
            ),
            MaterialDesignGridRow(
              children: [
                MaterialDesignGridItem(
                  columns: 4,
                  child: TextFormField(
                    minLines: 4,
                    maxLines: 4,
                    decoration: const InputDecoration(
                      labelText: 'Message*',
                      alignLabelWithHint: true,
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),
              ],
            ),
            const MaterialDesignGridRow(
              children: [
                MaterialDesignGridItem(
                  columns: 4,
                  child: Text(
                      'Iste aut cupiditate adipisci et iste incidunt est animi. Voluptatem natus reiciendis magni nesciunt magni. Reiciendis ad beatae. Odit corrupti eos voluptatem molestiae voluptatem qui rerum animi officiis. Natus dolores provident excepturi nostrum omnis voluptatem.'),
                ),
              ],
            ),
            MaterialDesignGridRow(
              children: [
                MaterialDesignGridItem(
                  columns: 4,
                  child: ElevatedButton(
                    onPressed: () {},
                    child: const Text('Submit'),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
